### PR TITLE
Knowledge Distillation support

### PR DESCRIPTION
**Pull request by tanglang96**
add Knowledge Distillation support and example

### CODE REVIEW

File: src/sdk/pynni/nni/compression/torch/knowledge_distill.py [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+        Returns
+        -------
+        torch.Tensor
+            weighted distillation loss
+        """
+
+        with torch.no_grad():
+            kd_out = self.teacher_model(data)
+        soft_log_out = F.log_softmax(student_out / self.kd_T, dim=1)
+        soft_t = F.softmax(kd_out / self.kd_T, dim=1)
```
Comment (chicm-ms) [inline src/sdk/pynni/nni/compression/torch/knowledge_distill.py]: It is very common that the output of a model is not  a tensor.  Here we require the model output to be a tensor, it is a limitation.

Comment (tanglang96) [reply src/sdk/pynni/nni/compression/torch/knowledge_distill.py]: If the output of the neural network is not a tensor, how can the loss and gradients be back-propagated?

Comment (chicm-ms) [reply src/sdk/pynni/nni/compression/torch/knowledge_distill.py]: the output can be a tuple or a list of tensors, such as object detection, or any multitask model.

Comment (tanglang96) [reply src/sdk/pynni/nni/compression/torch/knowledge_distill.py]: Thanks for the suggestion, I'll do some investigations

Comment (tanglang96) [reply src/sdk/pynni/nni/compression/torch/knowledge_distill.py]: I made some modifications to enable users to modify the teacher model's output(such as take the first tensor in a tuple), in this way they can decide which part to distill.

File: examples/model_compress/knowledge_distill/KnowledgeDistill.md [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+![](../../../docs/img/distill.png)
+
+### Usage
+
+PyTorch code
+
+```python
+from knowledge_distill.knowledge_distill import KnowledgeDistill
+kd = KnowledgeDistill(kd_teacher_model, kd_T=5, kd_beta=1)
+
```
Comment (chicm-ms) [inline examples/model_compress/knowledge_distill/KnowledgeDistill.md]: It is hard to understand this kd_beta without reading source code of KnowledgeDistill class. and considering the `student_loss` is passed to KnowledgeDistill only for add, my suggestion:
Remove the kd_beta and student_loss parameter of KnowledgeDistill.loss,   KnowledgeDistill is responsible only the KD loss, so that we can use it like this:
```
student_loss = F.cross_entropy(output, target)
kd_loss = kd.loss(data, output)
loss = student_loss * 0.2 + kd_loss * 0.8
...
```

Comment (tanglang96) [reply examples/model_compress/knowledge_distill/KnowledgeDistill.md]: resolved

File: examples/model_compress/knowledge_distill/knowledge_distill.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+class KnowledgeDistill():
+    """
+    Knowledge Distillaion support while fine-tuning the compressed model
+    Geoffrey Hinton, Oriol Vinyals, Jeff Dean
+    "Distilling the Knowledge in a Neural Network"
+    https://arxiv.org/abs/1503.02531
+    """
+
+    def __init__(self, teacher_model, kd_T):
```
Comment (chicm-ms) [inline examples/model_compress/knowledge_distill/knowledge_distill.py]: better to have a default value of 1

File: examples/model_compress/knowledge_distill/KnowledgeDistill.md [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+    loss = F.cross_entropy(output, target)
+    # you only to add the following line to fine-tune with knowledge distillation
+    loss = alpha * loss + beta * kd.loss(data=data, student_out=output)
+    loss.backward()
+```
+
+#### User configuration for Level Pruner
+* **kd_teacher_model:** The pre-trained teacher model 
+* **kd_T:** Temperature for smoothing teacher model's output
+* **kd_beta:** Weight of knowledge distillation loss relative to the training loss
```
Comment (chicm-ms) [inline examples/model_compress/knowledge_distill/KnowledgeDistill.md]: kd_beta can be removed.

File: examples/model_compress/knowledge_distill/KnowledgeDistill.md [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+    data, target = data.to(device), target.to(device)
+    optimizer.zero_grad()
+    output = model(data)
+    loss = F.cross_entropy(output, target)
+    # you only to add the following line to fine-tune with knowledge distillation
+    loss = alpha * loss + beta * kd.loss(data=data, student_out=output)
+    loss.backward()
+```
+
+#### User configuration for Level Pruner
```
Comment (chicm-ms) [inline examples/model_compress/knowledge_distill/KnowledgeDistill.md]: configuration for KD or Level pruner?

Comment (tanglang96) [reply examples/model_compress/knowledge_distill/KnowledgeDistill.md]: resolved.

File: examples/model_compress/knowledge_distill/KnowledgeDistill.md (hunk 4 above)
Comment (tanglang96) [reply examples/model_compress/knowledge_distill/KnowledgeDistill.md]: resolved.

File: examples/model_compress/knowledge_distill/knowledge_distill.py (hunk 3 above)
Comment (tanglang96) [reply examples/model_compress/knowledge_distill/knowledge_distill.py]: resolved.

File: examples/model_compress/knowledge_distill/KnowledgeDistill.md [hunk 6]
```diff
@@ -0,0 +1,31 @@
+Knowledge Distillation on NNI Compressor
```
Comment (QuanluZhang) [inline examples/model_compress/knowledge_distill/KnowledgeDistill.md]: not suggest to put doc under examples folder

File: examples/model_compress/knowledge_distill/KnowledgeDistill.md [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+===
+
+## KnowledgeDistill
+
+Knowledge distillation support while fine-tuning the compressed model, in [Distilling the Knowledge in a Neural Network](https://arxiv.org/abs/1503.02531),  the compressed model is trained to mimic a pre-trained, larger model.  This training setting is also referred to as "teacher-student",  where the large model is the teacher and the small model is the student.
```
Comment (QuanluZhang) [inline examples/model_compress/knowledge_distill/KnowledgeDistill.md]: `Knowledge distillation support while fine-tuning the compressed model`, don't understand...

File: examples/model_compress/knowledge_distill/KnowledgeDistill.md (hunk 6 above)
Comment (QuanluZhang) [reply examples/model_compress/knowledge_distill/KnowledgeDistill.md]: you could put it under `docs/en_US/TrialExample/`, and properly reference it in `Compressor/Overview.md`

**Comment by QuanluZhang** [comment]
the code looks great.

File: examples/model_compress/knowledge_distill/KnowledgeDistill.md (hunk 6 above)
Comment (tanglang96) [reply examples/model_compress/knowledge_distill/KnowledgeDistill.md]: resolved.

File: examples/model_compress/knowledge_distill/KnowledgeDistill.md (hunk 7 above)
Comment (tanglang96) [reply examples/model_compress/knowledge_distill/KnowledgeDistill.md]: resolved.
